File: Patterns/Builder_1.py
from typing import List, Dict, Optional


class Resume:
    """Класс резюме, создаваемый через строителя"""
    def __init__(
        self,
        name: str,
        email: str,
        phone: str = "",
        summary: str = "",
        experience: list[dict] = None,
        education: list[dict] = None,
        skills: list[str] = None,
        languages: list[str] = None,
        certifications: list[str] = None,
    ):
        self.name = name
        self.email = email
        self.phone = phone
        self.summary = summary
        self.experience = experience or []
        self.education = education or []
        self.skills = skills or []
        self.languages = languages or []
        self.certifications = certifications or []

    def __str__(self) -> str:
        result = [
            f"\n{'='*50}",
            f"=== {self.name} ===",
            f"Email: {self.email}",
            f"Телефон: {self.phone}" if self.phone else "",
            f"О себе: {self.summary}" if self.summary else "",
            f"\n--- Навыки ---",
            f"{', '.join(self.skills)}" if self.skills else "Не указаны",
            f"\n--- Опыт работы ---",
        ]
        for exp in self.experience:
            result.append(f"  • {exp.get('company', '')} — {exp.get('years', '')} лет: {exp.get('position', '')}")
        
        result.append("\n--- Образование ---")
        for edu in self.education:
            result.append(f"  • {edu.get('degree', '')} — {edu.get('school', '')}")
        
        if self.languages:
            result.append(f"\n--- Языки ---\n{', '.join(self.languages)}")
        if self.certifications:
            result.append(f"\n--- Сертификаты ---\n{', '.join(self.certifications)}")
        
        return "\n".join(filter(None, result))


class ResumeBuilder:
    """Строитель для пошагового создания резюме"""
    def __init__(self):
        self.reset()
    
    def reset(self):
        self._name = ""
        self._email = ""
        self._phone = ""
        self._summary = ""
        self._experience = []
        self._education = []
        self._skills = []
        self._languages = []
        self._certifications = []
        return self
    
    def set_name(self, name: str):
        self._name = name
        return self
    
    def set_contacts(self, email: str, phone: str = ""):
        self._email = email
        self._phone = phone
        return self
    
    def set_summary(self, summary: str):
        self._summary = summary
        return self
    
    def add_experience(self, company: str, years: int, position: str = ""):
        self._experience.append({"company": company, "years": years, "position": position})
        return self
    
    def add_education(self, degree: str, school: str, year: int = 0):
        self._education.append({"degree": degree, "school": school, "year": year})
        return self
    
    def add_skill(self, skill: str):
        self._skills.append(skill)
        return self
    
    def add_language(self, language: str):
        self._languages.append(language)
        return self
    
    def add_certification(self, certification: str):
        self._certifications.append(certification)
        return self
    
    def build(self) -> Resume:
        """Возвращает готовый объект Resume"""
        return Resume(
            name=self._name,
            email=self._email,
            phone=self._phone,
            summary=self._summary,
            experience=self._experience,
            education=self._education,
            skills=self._skills,
            languages=self._languages,
            certifications=self._certifications,
        )


class ResumeDirector:
    """Директор для создания стандартных конфигураций резюме"""
    
    @staticmethod
    def build_standard_resume(builder: ResumeBuilder, name: str, email: str) -> Resume:
        """Стандартное резюме: имя, контакты, базовые навыки"""
        return (builder
            .reset()
            .set_name(name)
            .set_contacts(email)
            .add_skill("Коммуникабельность")
            .add_skill("Работа в команде")
            .add_skill("Ответственность")
            .build()
        )
    
    @staticmethod
    def build_extended_resume(
        builder: ResumeBuilder, 
        name: str, 
        email: str, 
        phone: str,
        company: str,
        years: int,
        degree: str,
        school: str
    ) -> Resume:
        """Расширенное резюме: имя, контакты, опыт, образование, навыки, языки, сертификаты"""
        return (builder
            .reset()
            .set_name(name)
            .set_contacts(email, phone)
            .set_summary(f"Опытный специалист с {years}+ годами опыта в {company}")
            .add_experience(company, years, "Senior Developer")
            .add_education(degree, school, 2020)
            .add_skill("Python")
            .add_skill("SQL")
            .add_skill("Docker")
            .add_skill("Git")
            .add_language("Английский (B2)")
            .add_language("Немецкий (A2)")
            .add_certification("AWS Certified Developer")
            .add_certification("Kubernetes Basics")
            .build()
        )

File: Patterns/Builder_1.php
<?php

// Класс Resume
class Resume {
    public function __construct(
        public readonly string $name,
        public readonly string $email,
        public readonly string $phone = '',
        public readonly string $summary = '',
        public readonly array  $experience = [],
        public readonly array  $education = [],
        public readonly array  $skills = [],
        public readonly array  $languages = [],
        public readonly array  $certifications = [],
    ) {}

    public function __toString(): string {
        return sprintf(
            "=== %s ===\nEmail: %s\nНавыки: %s\nОпыт: %d позиций",
            $this->name,
            $this->email,
            implode(', ', $this->skills),
            count($this->experience)
        );
    }
}

// Строитель для резюме
class ResumeBuilder {
    private string $name = '';
    private string $email = '';
    private string $phone = '';
    private string $summary = '';
    private array $experience = [];
    private array $education = [];
    private array $skills = [];
    private array $languages = [];
    private array $certifications = [];
    
    public function set_name(string $name): self {
        $this->name = $name;
        return $this;
    }
    
    public function set_contacts(string $email, string $phone = ''): self {
        $this->email = $email;
        $this->phone = $phone;
        return $this;
    }
    
    public function add_experience(string $company, int $years, string $position = ''): self {
        $this->experience[] = [
            'company' => $company,
            'years' => $years,
            'position' => $position
        ];
        return $this;
    }
    
    public function add_education(string $degree, string $school, int $year = 0): self {
        $this->education[] = [
            'degree' => $degree,
            'school' => $school,
            'year' => $year
        ];
        return $this;
    }
    
    public function add_skill(string $skill): self {
        $this->skills[] = $skill;
        return $this;
    }
    
    public function add_language(string $language, string $level = ''): self {
        $this->languages[] = [
            'language' => $language,
            'level' => $level
        ];
        return $this;
    }
    
    public function add_certification(string $certification): self {
        $this->certifications[] = $certification;
        return $this;
    }
    
    public function set_summary(string $summary): self {
        $this->summary = $summary;
        return $this;
    }
    
    public function build(): Resume {
        // Проверка обязательных полей
        if (empty($this->name)) {
            throw new Exception("Имя обязательно для резюме");
        }
        if (empty($this->email)) {
            throw new Exception("Email обязателен для резюме");
        }
        
        return new Resume(
            name: $this->name,
            email: $this->email,
            phone: $this->phone,
            summary: $this->summary,
            experience: $this->experience,
            education: $this->education,
            skills: $this->skills,
            languages: $this->languages,
            certifications: $this->certifications
        );
    }
}

// Директор для сборки стандартных резюме
class Director {
    private ResumeBuilder $builder;
    
    public function __construct(ResumeBuilder $builder) {
        $this->builder = $builder;
    }
    
    // Стандартное резюме
    public function buildStandardResume(string $name, string $email): Resume {
        return $this->builder
            ->set_name($name)
            ->set_contacts($email)
            ->add_skill('Коммуникабельность')
            ->add_skill('Работа в команде')
            ->add_skill('Ответственность')
            ->set_summary('Молодой специалист с желанием развиваться')
            ->build();
    }
    
    // Расширенное резюме (с опытом и образованием)
    public function buildExtendedResume(
        string $name, 
        string $email, 
        string $phone,
        string $summary
    ): Resume {
        return $this->builder
            ->set_name($name)
            ->set_contacts($email, $phone)
            ->set_summary($summary)
            ->add_skill('Python')
            ->add_skill('SQL')
            ->add_skill('Docker')
            ->add_skill('Git')
            ->add_experience('Яндекс', 3, 'Разработчик')
            ->add_experience('Google', 1, 'Стажёр')
            ->add_education('Бакалавр', 'МГУ', 2020)
            ->add_education('Магистр', 'МФТИ', 2022)
            ->add_language('Английский', 'B2')
            ->add_language('Немецкий', 'A2')
            ->add_certification('AWS Practitioner')
            ->build();
    }
}

// Клиентский код

echo "=== Стандартное резюме ===\n";
$builder = new ResumeBuilder();
$director = new Director($builder);
$standardResume = $director->buildStandardResume('Анна Иванова', 'anna@example.com');
echo $standardResume . "\n\n";

echo "=== Расширенное резюме ===\n";
$builder2 = new ResumeBuilder();
$director2 = new Director($builder2);
$extendedResume = $director2->buildExtendedResume(
    'Петр Сидоров',
    'petr@example.com',
    '+7-900-123-45-67',
    'Опытный разработчик с 5-летним стажем'
);
echo $extendedResume . "\n\n";

echo "=== Ручное построение резюме через билдер ===\n";
$customResume = (new ResumeBuilder())
    ->set_name('Мария Петрова')
    ->set_contacts('maria@example.com', '+7-901-234-56-78')
    ->add_experience('Сбербанк', 2, 'Аналитик')
    ->add_education('Бакалавр', 'ВШЭ', 2019)
    ->add_skill('SQL')
    ->add_skill('Tableau')
    ->add_skill('Excel')
    ->add_language('Английский', 'C1')
    ->set_summary('Аналитик данных с опытом работы в крупной компании')
    ->build();

echo $customResume . "\n\n";

// Проверка валидации
echo "=== Проверка валидации ===\n";
try {
    $invalidResume = (new ResumeBuilder())
        ->set_name('Тест')
        ->build(); // Нет email
} catch (Exception $e) {
    echo "Ошибка: " . $e->getMessage() . "\n";
}